src/tasks/withdraw.ts

```diff
@@ -173,12 +173,56 @@ async function appraise(
   return { ...token, usdValue, decimals };
 }
 
-async function getAllTradedTokens(settlement: Contract): Promise<string[]> {
-  const trades = await settlement.queryFilter(settlement.filters.Trade());
-  const tokens = new Set(
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    trades.map((trade) => [trade.args!.sellToken, trade.args!.buyToken]).flat(),
-  );
+function isErrorTooManyEvents(error: Error): boolean {
+  return /query returned more than \d* results/.test(error.message);
+}
+
+// List all traded tokens. Block range bounds are both inclusive.
+async function getAllTradedTokens(
+  settlement: Contract,
+  fromBlock: number,
+  toBlock: number,
+  hre: HardhatRuntimeEnvironment,
+): Promise<string[]> {
```

Comment:
I'm not sure why we are bound to looking at trade events for tokens. Why not just use whatever the safe used to display your token balances. This would likely require significantly fewer queries. I mean there are at least three apps that can display a single wallets asset status (Zapper, Zerion, Gnosis Safe). Have we considered this route?

Reply:
On the one hand, I like that the script has no reliance on external services - you just need a node. On the other hand, I do like the suggestion of using an external service to get token balances as that would greatly speed up the process of figuring out all traded tokens.

Since we were already using trade events for this, I think its fine for this PR to continue using traded events and just fix the issue of querying too many events at once, but we should create an issue and consider exploring @bh2smith's suggestion of using an external service to get this data (or at least allow it in some way).

We could also use a subgraph :smile:.

Reply:
As a motivation why it's like this: In general I like to use as few services as possible so that the script doesn't stop working if one of the many services we rely on is unreachable/stops working. I kind of expect that services might require auth tokens like Infura, which would add friction to using the script. I also fear rate limits: since we have about 1100 tokens this script is currently processing 5 tokens/second.
But I agree at this point it makes sense using other services for retrieving balances, I'll look around for subgraphs, but even just a token list would be quite an improvement. :slightly_smiling_face: